include/jwt-cpp/jwt.h

```diff
@@ -2668,6 +2671,170 @@ namespace jwt {
+		 */
+		typename json_traits::string_type get_algorithm() const { return get_jwk_claim("alg").as_string(); }
+
+		/**
+		 * Get key id claim
+		 * \return key id as string
+		 * \throw std::runtime_error If claim was not present
+		 * \throw std::bad_cast Claim was present but not a string (Should not happen in a valid token)
+		 */
+		typename json_traits::string_type get_key_id() const { return get_jwk_claim("kid").as_string(); }
+
+		/**
+		 * Get x5c claim
+		 * \return x5c as string
+		 * \throw std::runtime_error If claim was not present
+		 * \throw std::bad_cast Claim was present but not a string (Should not happen in a valid token)
+		 */
+		typename json_traits::string_type get_x5c() const { return json_traits::as_string(get_jwk_claim("x5c").as_array().front()); } //do not use serialize() instead
+
+		/**
+		 * Check if algortihm is present ("alg")
+		 * \return true if present, false otherwise
+		 */
+		bool has_algorithm() const noexcept { return has_jwk_claim("alg"); }
+
+		/**
+		 * Check if key id is present ("kid")
+		 * \return true if present, false otherwise
+		 */
+		bool has_key_id() const noexcept { return has_jwk_claim("kid"); }
+
+		/**
+		 * Check if x5c is present ("x5c")
+		 * \return true if present, false otherwise
+		 */
+		bool has_x5c() const noexcept { return has_jwk_claim("x5c"); }
+
+		/**
+		 * Check if a jwks claim is present
+		 * \return true if claim was present, false otherwise
+		 */
+		bool has_jwk_claim(const typename json_traits::string_type& name) const noexcept { return jwk_claims.count(name) != 0; }
+
+		/**
+		 * Get jwks claim
+		 * \return Requested claim
+		 * \throw std::runtime_error If claim was not present
+		 */
+		basic_claim_t get_jwk_claim(const typename json_traits::string_type& name) const {
+			if (!has_jwk_claim(name)) {
+				throw std::runtime_error("claim not found");
+			}
+			return jwk_claims.at(name);
+		}
+
+		bool empty() const noexcept { return jwk_claims.empty(); }
+
+	private:
+		std::unordered_map<typename json_traits::string_type, basic_claim_t> jwk_claims;
+	};
+
+	template<typename json_traits>
+	class jwks {
+		using basic_claim_t = basic_claim<json_traits>;
+		using jwk_t = jwk<json_traits>;
+
+	public:
+
+		JWT_CLAIM_EXPLICIT jwks(const typename json_traits::string_type& jwks_token) {
+			auto parse_claims = [](const typename json_traits::string_type& str) {
+				std::unordered_map<typename json_traits::string_type, jwk_t> res;
+				typename json_traits::value_type val;
+				if (!json_traits::parse(val, str))
+					throw std::runtime_error("Invalid json");
+
+				for (const auto& k : json_traits::as_object(val)) {
+					size_t id = 0;
+					for (const typename json_traits::value_type& item : json_traits::as_array(k.second)) {
+						jwk_t jwk_entry(item);
+						if (jwk_entry.has_key_id())
+							res.emplace(jwk_entry.get_key_id(), std::move(jwk_entry));
+						else
+							res.emplace(std::to_string(id), std::move(jwk_entry));
+						++id;
+					}
+				}
```

Comment:
Yep, so I would really like if we could leave the KID outside of the JWKS logic because it's important to support workflows that do not required them. 

I will come back with some changes against your branch, I think the biggest limitation is the parsing or "map of claims"

I also see we don't check `"keys"` explicitly is the JWKS which would open use to a vulnerability